library/core/src/convert/mod.rs

```diff
@@ -93,6 +93,8 @@ pub use num::FloatToInt;
 ///
 /// let iter = [Some(1), None, Some(3)].into_iter();
 /// let filtered = iter.filter_map(identity).collect::<Vec<_>>();
+/// // Equivalent, since Option is also a Iterator itself:
+/// // let filtered = iter.flatten().collect::<Vec<_>>();
```

Comment:
`Option` is _not_ an `Iterator`, but it does implement `IntoIterator`, which is sufficient for `flatten()`.

Although `flatten()` is shorter to write, I personally still prefer `filter_map(identity)` because it is a much simpler iterator, and simple is usually better for optimization.

Reply:
I hadn't really thought about the optimization aspect, just that I prefer not to import more libraries. That makes the advice previously given in the docs not as bad as I thought.

Reply:
Yeah, unfortunately `.flatten()` is a `flat_map`, not a `filter_map`, and thus fundamentally more complex.  `FlatMap` is bigger than `FilterMap` because it needs to store the prefix and suffix iterators, as those might have more elements in them, whereas because `FilterMap` only deals with `Option`s so knows they can't have another item in them after getting the one value out.  (You can also see this by how, for example, `filter_map` preserves the upper bound on a `size_hint`, which `flat_map` doesn't, though admittedly the upper bounds on size hints are useless.)  And this is why https://rust-lang.github.io/rust-clippy/master/index.html#flat_map_option exists.

I see that `flat_map` mentions `flatten` in the prose, but doesn't have an example -- maybe it'd make sense to move an example like this (perhaps with arrays instead of options) over to that method?

Reply:
(Also, personally I'd never use `identity` here, since it needs a `use` and even ignoring that it isn't even shorter than writing out the `|x| x` closure.  But the example using it isn't necessarily *wrong*.)

Reply:
That makes sense. Thank you for your insight, I especially appreciate including references and verifiable facts. I'll try to improve the documentation somewhere else.

I'll close this pull request.

Reply:
Also, I'm glad you raised this, since it made me look for a `performance` lint in clippy talking about this, but instead I found exactly the opposite, a lint *suggesting* the `.filter_map(identity)` -> `.flatten()` change.

So I've filed https://github.com/rust-lang/rust-clippy/issues/9377 for the clippy folks to discuss the situation.